src/shim/preload_libraries.c

```diff
@@ -477,3 +480,70 @@ FILE *tmpfile(void) {
         free(name);
     return rv;
 }
+
+static void _convert_stat_to_stat64(struct stat* s, struct stat64* s64) {
+    memset(s64, 0, sizeof(*s64));
+
+    s64->st_dev = s->st_dev;
+    s64->st_ino = s->st_ino;
+    s64->st_nlink = s->st_nlink;
+    s64->st_mode = s->st_mode;
+    s64->st_uid = s->st_uid;
+    s64->st_gid = s->st_gid;
+    s64->st_rdev = s->st_rdev;
+    s64->st_size = s->st_size;
+    s64->st_blksize = s->st_blksize;
+    s64->st_blocks = s->st_blocks;
+    s64->st_atim = s->st_atim;
+    s64->st_mtim = s->st_mtim;
+    s64->st_ctim = s->st_ctim;
+}
+
+static void _convert_statfs_to_statfs64(struct statfs* s, struct statfs64* s64) {
+    memset(s64, 0, sizeof(*s64));
+
+    s64->f_type = s->f_type;
+    s64->f_bsize = s->f_bsize;
+    s64->f_blocks = s->f_blocks;
+    s64->f_bfree = s->f_bfree;
+    s64->f_bavail = s->f_bavail;
+    s64->f_files = s->f_files;
+    s64->f_ffree = s->f_ffree;
+    s64->f_fsid = s->f_fsid;
+    s64->f_namelen = s->f_namelen;
+    s64->f_frsize = s->f_frsize;
+    s64->f_flags = s->f_flags;
+}
+
+// Some platforms define fstat and fstatfs as macros. We should call 'syscall()' directly since
+// calling for example 'fstat()' will not necessarily call shadow's 'fstat()' wrapper defined in
+// 'preload_syscalls.c'.
+
+int fstat64(int a, struct stat64* b) {
+    struct stat s;
+    int rv = syscall(SYS_fstat, a, &s);
+    _convert_stat_to_stat64(&s, b);
+    return rv;
+}
+
+int fstatfs64(int a, struct statfs64* b) {
+    struct statfs s;
+    int rv = syscall(SYS_fstatfs, a, &s);
+    _convert_statfs_to_statfs64(&s, b);
+    return rv;
+}
+
+int __fxstat(int ver, int a, struct stat* b) {
+    // the version doesn't matter because we never make the syscall directly,
+    // and will later pass it back to glibc's fstat
```

Comment:
By 'pass it back to', do you mean we expect this to only ever be called by glibc's fstat?

Reply:
I worded this badly. See above.